doc/dev/adapters.md

```diff
@@ -1,15 +1,699 @@
+# Istio Mixer: Adapter Developer's Guide
 
-This will eventually turn into a developer's guide for 
-creating Mixer adapters. For now, it's just a set of
-notes and reminders:
+This document is or developers looking to build an [adapter](https://istio.io/docs/concepts/policy-and-control/mixer.html#adapters)
+for Istio's Mixer. Adapters integrate Mixer with different infrastructure backends that deliver core functionality, such
+as logging, monitoring, quotas, ACL checking, and more. This guide explains the adapter model and adapter lifecycle, and
+also walks through the step-by-step instructions for creating a simple adapter.
 
-- Adapters must use env.Logger for logging during
-execution. This logger understands about which adapter
-is running and routes the data to the place where the
-operator wants to see it.
+**Prerequisite**: This developer guide assumes the reader has:
+
+* Basic understanding of the [Mixer Framework](https://istio.io/docs/concepts/policy-and-control/mixer.html) as an attribute processing engine.
+
+* High level understanding of [operator configuration model](https://istio.io/docs/concepts/policy-and-control/mixer-config.html).
+
+# Table of contents
+* [Background](#Background)
+* [Template overview](#template-overview)
+* [Adapter life cycle](#adapter-life-cycle)
+* [Example](#example)
+* [Summary diagrams](#summary-diagrams)
+* [Plug adapter into Mixer](#plug-adapter-into-Mixer)
+* [Testing](#testing)
+* [Do's and dont's](#dos-and-donts)
+* [Built-in templates](#built-in-templates)
+* [Walkthrough](#walkthrough-of-adapter-implementation-30-minutes)
+
+# Background
+
+[Mixer](https://istio.io/docs/concepts/policy-and-control/mixer.html) provides Istio's generic intermediation layer
+between application code and infrastructure backends such as billing systems, access control systems, logging, and so
+on. It is an [attribute](https://istio.io/docs/concepts/policy-and-control/attributes.html) processing engine that uses
+operator-supplied [configuration](https://istio.io/docs/concepts/policy-and-control/mixer-config.html) to map request
+attributes from proxy into calls to these backend systems via a pluggable set of [adapters](https://istio.io/docs/concepts/policy-and-control/mixer.html#adapters).
+Adapters enable Mixer to expose a single consistent API, independent of the infrastructure backends in use. The exact
+set of adapters used at runtime is determined through operator configuration and can easily be extended to target new
+or custom infrastructure backends.
+
+![mixer architecture](./img/mixer%20architecture.svg)
+
+## Templates, Adapters and Operator Configuration
+
+Mixer structures its incoming attribute data into a more useful form for backends (for example, metrics data) using
+templates. Templates describe the form of data dispatched to adapters when processing a request and the interface that
+the adapter must implement to receive this data. Mixer provides a range of default templates suitable for most use
```

Comment:
Simplify:

Out of the box, Mixer provides a suite of default templates. We strongly recommend...

And make "default templates" into a link